lib/hooks.js

```diff
@@ -181,416 +164,27 @@ const Hooks = {
     hookType = getProxiedHooks(hookType);
 
     for (const type of hookType) {
-      this.options.hooks[type] = this.options.hooks[type].filter(hook => {
-        if (isReference && typeof hook === 'function') {
-          return hook !== name; // check if same method
-        }
-        if (!isReference && typeof hook === 'object') {
-          return hook.name !== name;
-        }
-        return true;
-      });
+      this.hooks[type] = this.hooks[type].filter(hookFn => fn !== hookFn);
     }
 
     return this;
-  },
+  }
 
   /**
    * Check whether the mode has any hooks of this type
    *
    * @param {string} hookType @see {@link hookTypes}
-   *
-   * @alias hasHooks
-   *
-   * @memberof Sequelize
-   * @memberof Sequelize.Model
    */
-  hasHook(hookType) {
-    return this.options.hooks[hookType] && !!this.options.hooks[hookType].length;
+  has(hookType) {
+    return this.hooks[hookType] && !!this.hooks[hookType].length;
   }
-};
-Hooks.hasHooks = Hooks.hasHook;
-
 
-function applyTo(target, isModel = false) {
-  _.mixin(target, Hooks);
-
-  for (const hook of Object.keys(hookTypes)) {
-    if (isModel && hookTypes[hook].noModel) {
-      continue;
-    }
-    target[hook] = function(name, callback) {
-      return this.addHook(hook, name, callback);
-    };
+  /**
+   * Removes all hooks
+   */
+  removeAll() {
+    this.hooks = {};
   }
 }
-exports.applyTo = applyTo;
-
-/**
- * A hook that is run before validation
- *
- * @param {string}   name
- * @param {Function} fn   A callback function that is called with instance, options
- * @name beforeValidate
```

Comment:
We should preserve / update this documentation. It can be placed alongside respective `hooksType` entry

Reply:
Can do, while I'm there, do you mind if we remove the `params: <NUMBER>` bit?
I just noticed this data is not used at all.

Reply:
`param` indicate how many arguments are passed to hook, so you can deal with those based on that data.

Reply:
We can remove `param` if documentation is kept with hook definition, thus clearly stating what arguments are being passed

Reply:
Will do, however expressing that 'clearly' is rather difficult with the limitation of jsdoc.
The typings will provide the most accurate API documentation here.

Reply:
Just state what kind of object is being passed, jsdoc can handle that very well. Old docs from this file can be a starting point. Not everyone is using Typescript so we can't treat them as documentation

Reply:
@sushantdhiman I did recover the documentation, probably not like you envisioned it though.

Feel free to give me an example because I really don't use jsdoc and have no idea how to model this accordingly.

Reply:
```js
/**
 * A hook that is run after validation
 *
 * @param {string}   name
 * @param {Function} fn   A callback function that is called with instance, options
 * @name afterValidate
 * @memberof Sequelize.Model
 */
```

This is want I was thinking